### PR TITLE
Compile-only API dependencies can be declared for JVM libraries

### PR DESCRIPTION
**Pull request by jjohannes**
When writing a Java (or Groovy/Kotlin/Scala) library, there are cases where you require dependencies at compilation time which are parts of the API of your library, but which should not be on the runtime classpath.

An example of such a dependency is an annotation library with annotations that are not used at runtime. These typically need to be available at compile time of the library's consumers when annotation processors inspect annotations of all classes. Another example is a dependency that is part of the runtime environment the library is expected to run on, but also provides types that are used in the public API of the library.

This PR lets the Java Library add the `compileOnlyApi` configuration.

(This is also a follow up to #14240 where `compileOnlyApi` is already used)

### CODE REVIEW

**Comment by jjohannes** [comment]
@bot-gradle test this

File: subprojects/docs/src/docs/userguide/jvm/building_java_projects.adoc [hunk 1]
```diff
@@ -150,7 +150,7 @@ As far as configurations go, the main ones of interest are:
 
 You can learn more about these and how they relate to one another in the <<java_plugin.adoc#sec:java_plugin_and_dependency_management,plugin reference chapter>>.
 
-Be aware that the <<java_library_plugin.adoc#java_library_plugin,Java Library Plugin>> creates an additional configuration — `api` — for dependencies that are required for compiling both the module and any modules that depend on it.
+Be aware that the <<java_library_plugin.adoc#java_library_plugin,Java Library Plugin>> offers two additional configuration — `api` and `compileOnlyApi` — for dependencies that are required for compiling both the module and any modules that depend on it.
```
Comment (melix) [inline subprojects/docs/src/docs/userguide/jvm/building_java_projects.adoc:153]:
```suggestion
Be aware that the <<java_library_plugin.adoc#java_library_plugin,Java Library Plugin>> offers two additional configurations — `api` and `compileOnlyApi` — for dependencies that are required for compiling both the module and any modules that depend on it.
```

File: subprojects/docs/src/docs/userguide/jvm/java_library_plugin.adoc [hunk 2]
```diff
@@ -123,7 +123,7 @@ include::sample[dir="snippets/java-library/quickstart/kotlin",files="build.gradl
 [[sec:java_library_configurations_graph]]
 == The Java Library plugin configurations
 
-The following graph describes the main configurations setup when the Java Library plugin is in use.
+The following graph describes the configurations setup when the Java Library plugin is in use.
```
Comment (melix) [inline subprojects/docs/src/docs/userguide/jvm/java_library_plugin.adoc:126]:
```suggestion
The following graph describes how configurations are setup when the Java Library plugin is in use.
```

**Review by melix — APPROVED** [review]
LGTM with minor suggestions for docs

**Pull request merged**